### ISSUE TITLE
Optional[T] of T never match `None`

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#609]
I want to define the following `unwrap` function, which unwrap `T` from `Optional[T]`.

```python
from typing import Optional, TypeVar
T = TypeVar("T")
def unwrap(v: Optional[T]) -> T:
    if v is None:
        raise ValueError
    return v
```

However this `unwrap` doesn't work well with pyright, while it works well with mypy. 

```python
import random
def f() -> Optional[int]:
    return random.choice([0, None])
a: int = unwrap(f())  # Error: Type "None" cannot be assigned to type "int"
```

I think the cause is that pyright resolves `T` with `Optional[int]` instead of `int`.

I'm glad if pyright works well with this situation.

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem, and I'll look into it.

**Comment by erictraut** [comment]
This will be fixed in the next published version of pyright. Thanks again for the bug report!

**Comment by erictraut** [comment]
This is fixed in 1.1.31, which I just published.
